stdlib/list.gr

```diff
@@ -548,6 +548,77 @@ export let unique = list => {
   iter(list, [])
 }
 
+/**
+ * Produces a new list filled with tuples of elements from both given lists.
+ * The first tuple will contain the first item of each list, the second tuple
+ * will contain the second item of each list, and so on.
+ * 
+ * Calling this function with lists of different sizes will cause the returned
+ * list to have the length of the smaller list.
+ *
+ * @param list1: The list to provide values for the first tuple element
+ * @param list2: The list to provide values for the second tuple element
+ * @returns The new list containing indexed pairs of `(a, b)`
+ *
+ * @example List.zip([1, 2, 3], [4, 5, 6]) // [(1, 4), (2, 5), (3, 6)]
+ * @example List.zip([1, 2, 3], [4, 5]) // [(1, 4), (2, 5)]
+ * 
+ * @since v0.5.3
+ */
+export let rec zip = (list1, list2) => {
+  match ((list1, list2)) {
+    ([], []) => [],
+    ([first1, ...rest1], [first2, ...rest2]) => [(first1, first2), ...zip(rest1, rest2)],
+    _ => []
+  }
```

Comment:
I think this needs to use a helper function to be made tail-recursive. I believe you can look at `List.length` or `List.reverse` to see the pattern for doing that.

Reply:
Would something like this work?
```
let zip = (list1, list2) => {
  let rec zipInner = (list1, list2, acc) => {
    match ((list1, list2)) {
      ([first1, ...rest1], [first2, ...rest2]) => zipInner(rest1, rest2, [(first1, first2), ...acc]),
      _ => reverse(acc)
    }
  }
  zipInner(list1, list2, [])
}
```
I add the next elements in the list to the beginning of the accumulator and then reverse it at the end to bring it back to its normal shape